File: infra/src/main/java/com/beat/infra/persistence/booking/repository/query/MakerTicketReadPortImpl.java
package com.beat.infra.persistence.booking.repository.query;

import java.util.List;

import org.springframework.stereotype.Repository;

import com.beat.contracts.booking.MakerTicketReadPort;
import com.beat.contracts.booking.readmodel.MakerTicketBookingStatus;
import com.beat.contracts.booking.readmodel.MakerTicketListItemReadModel;
import com.beat.contracts.booking.readmodel.MakerTicketScheduleNumber;
import com.beat.domain.booking.domain.BookingStatus;
import com.beat.domain.performance.domain.BankName;
import com.beat.domain.schedule.domain.ScheduleNumber;
import com.beat.infra.persistence.booking.entity.BookingJpaEntity;

import jakarta.persistence.EntityManager;
import jakarta.persistence.TypedQuery;

@Repository
public class MakerTicketReadPortImpl implements MakerTicketReadPort {

	private final EntityManager entityManager;

	public MakerTicketReadPortImpl(EntityManager entityManager) {
		this.entityManager = entityManager;
	}

	@Override
	public List<MakerTicketListItemReadModel> findTickets(Long performanceId,
		List<MakerTicketScheduleNumber> scheduleNumberNames, List<MakerTicketBookingStatus> bookingStatusNames) {
		List<ScheduleNumber> scheduleNumbers = toScheduleNumbers(scheduleNumberNames);
		List<BookingStatus> bookingStatuses = toBookingStatuses(bookingStatusNames);

		StringBuilder jpql = baseBookingScheduleJpql();
		appendPerformanceCondition(jpql, performanceId);
		appendScheduleNumberCondition(jpql, scheduleNumbers);
		appendBookingStatusCondition(jpql, bookingStatuses);
		jpql.append(orderByBookingStatusAndCreatedAt());

		TypedQuery<BookingJpaEntity> query = entityManager.createQuery(jpql.toString(), BookingJpaEntity.class);
		bindCommonParameters(query);
		bindPerformanceParameter(query, performanceId);
		bindScheduleNumberParameter(query, scheduleNumbers);
		bindBookingStatusParameter(query, bookingStatuses);
		return toReadModels(query.getResultList());
	}

	@Override
	public List<MakerTicketListItemReadModel> searchTickets(Long performanceId, String searchWord,
		List<MakerTicketScheduleNumber> scheduleNumberNames, List<MakerTicketBookingStatus> bookingStatusNames) {
		if (searchWord == null || searchWord.isBlank()) {
			return List.of();
		}

		List<ScheduleNumber> scheduleNumbers = toScheduleNumbers(scheduleNumberNames);
		List<BookingStatus> bookingStatuses = toBookingStatuses(bookingStatusNames);

		StringBuilder jpql = baseBookingScheduleJpql();
		appendPerformanceCondition(jpql, performanceId);
		appendScheduleNumberCondition(jpql, scheduleNumbers);
		appendBookingStatusCondition(jpql, bookingStatuses);
		jpql.append("  AND function('match', b.bookerName, :searchWord) > 0\n");
		jpql.append(orderByBookingStatusAndCreatedAt());

		TypedQuery<BookingJpaEntity> query = entityManager.createQuery(jpql.toString(), BookingJpaEntity.class);
		bindCommonParameters(query);
		bindPerformanceParameter(query, performanceId);
		bindScheduleNumberParameter(query, scheduleNumbers);
		bindBookingStatusParameter(query, bookingStatuses);
		query.setParameter("searchWord", searchWord);
		return toReadModels(query.getResultList());
	}

	private StringBuilder baseBookingScheduleJpql() {
		return new StringBuilder("""
			SELECT b
			FROM Booking b, Schedule s
			WHERE s.id = b.scheduleId
			  AND b.bookingStatus <> :deletedStatus
			""");
	}

	private void appendPerformanceCondition(StringBuilder jpql, Long performanceId) {
		if (performanceId != null) {
			jpql.append("  AND s.performanceId = :performanceId\n");
		}
	}

	private void appendScheduleNumberCondition(StringBuilder jpql, List<ScheduleNumber> scheduleNumbers) {
		if (scheduleNumbers != null && !scheduleNumbers.isEmpty()) {
			jpql.append("  AND s.scheduleNumber IN :scheduleNumbers\n");
		}
	}

	private void appendBookingStatusCondition(StringBuilder jpql, List<BookingStatus> bookingStatuses) {
		if (bookingStatuses != null && !bookingStatuses.isEmpty()) {
			jpql.append("  AND b.bookingStatus IN :bookingStatuses\n");
		}
	}

	private String orderByBookingStatusAndCreatedAt() {
		return """
			ORDER BY
			  CASE
			    WHEN b.bookingStatus = :refundRequestedStatus THEN 1
			    WHEN b.bookingStatus = :checkingPaymentStatus THEN 2
			    WHEN b.bookingStatus = :bookingConfirmedStatus THEN 3
			    WHEN b.bookingStatus = :bookingCancelledStatus THEN 4
			    ELSE 5
			  END ASC,
			  b.createdAt DESC
			""";
	}

	private void bindCommonParameters(TypedQuery<BookingJpaEntity> query) {
		query.setParameter("deletedStatus", BookingStatus.BOOKING_DELETED);
		query.setParameter("refundRequestedStatus", BookingStatus.REFUND_REQUESTED);
		query.setParameter("checkingPaymentStatus", BookingStatus.CHECKING_PAYMENT);
		query.setParameter("bookingConfirmedStatus", BookingStatus.BOOKING_CONFIRMED);
		query.setParameter("bookingCancelledStatus", BookingStatus.BOOKING_CANCELLED);
	}

	private void bindPerformanceParameter(TypedQuery<BookingJpaEntity> query, Long performanceId) {
		if (performanceId != null) {
			query.setParameter("performanceId", performanceId);
		}
	}

	private void bindScheduleNumberParameter(TypedQuery<BookingJpaEntity> query, List<ScheduleNumber> scheduleNumbers) {
		if (scheduleNumbers != null && !scheduleNumbers.isEmpty()) {
			query.setParameter("scheduleNumbers", scheduleNumbers);
		}
	}

	private void bindBookingStatusParameter(TypedQuery<BookingJpaEntity> query, List<BookingStatus> bookingStatuses) {
		if (bookingStatuses != null && !bookingStatuses.isEmpty()) {
			query.setParameter("bookingStatuses", bookingStatuses);
		}
	}

	private List<ScheduleNumber> toScheduleNumbers(List<MakerTicketScheduleNumber> scheduleNumbers) {
		if (scheduleNumbers == null || scheduleNumbers.isEmpty()) {
			return List.of();
		}
		return scheduleNumbers.stream()
			.map(this::toScheduleNumber)
			.toList();
	}

	private List<BookingStatus> toBookingStatuses(List<MakerTicketBookingStatus> bookingStatuses) {
		if (bookingStatuses == null || bookingStatuses.isEmpty()) {
			return List.of();
		}
		return bookingStatuses.stream()
			.map(this::toBookingStatus)
			.toList();
	}

	private ScheduleNumber toScheduleNumber(MakerTicketScheduleNumber scheduleNumber) {
		return switch (scheduleNumber) {
			case FIRST -> ScheduleNumber.FIRST;
			case SECOND -> ScheduleNumber.SECOND;
			case THIRD -> ScheduleNumber.THIRD;
			case FOURTH -> ScheduleNumber.FOURTH;
			case FIFTH -> ScheduleNumber.FIFTH;
			case SIXTH -> ScheduleNumber.SIXTH;
			case SEVENTH -> ScheduleNumber.SEVENTH;
			case EIGHTH -> ScheduleNumber.EIGHTH;
			case NINTH -> ScheduleNumber.NINTH;
			case TENTH -> ScheduleNumber.TENTH;
		};
	}

	private BookingStatus toBookingStatus(MakerTicketBookingStatus bookingStatus) {
		return switch (bookingStatus) {
			case CHECKING_PAYMENT -> BookingStatus.CHECKING_PAYMENT;
			case BOOKING_CONFIRMED -> BookingStatus.BOOKING_CONFIRMED;
			case BOOKING_CANCELLED -> BookingStatus.BOOKING_CANCELLED;
			case REFUND_REQUESTED -> BookingStatus.REFUND_REQUESTED;
			case BOOKING_DELETED -> BookingStatus.BOOKING_DELETED;
		};
	}

	private MakerTicketBookingStatus toMakerTicketBookingStatus(BookingStatus bookingStatus) {
		return switch (bookingStatus) {
			case CHECKING_PAYMENT -> MakerTicketBookingStatus.CHECKING_PAYMENT;
			case BOOKING_CONFIRMED -> MakerTicketBookingStatus.BOOKING_CONFIRMED;
			case BOOKING_CANCELLED -> MakerTicketBookingStatus.BOOKING_CANCELLED;
			case REFUND_REQUESTED -> MakerTicketBookingStatus.REFUND_REQUESTED;
			case BOOKING_DELETED -> MakerTicketBookingStatus.BOOKING_DELETED;
		};
	}

	private List<MakerTicketListItemReadModel> toReadModels(List<BookingJpaEntity> entities) {
		return entities.stream()
			.map(this::toReadModel)
			.toList();
	}

	private MakerTicketListItemReadModel toReadModel(BookingJpaEntity entity) {
		return new MakerTicketListItemReadModel(
			entity.getId(),
			entity.getBookerName(),
			entity.getBookerPhoneNumber(),
			entity.getScheduleId(),
			entity.getPurchaseTicketCount(),
			entity.getCreatedAt(),
			toMakerTicketBookingStatus(entity.getBookingStatus()),
			toBankName(entity.getBankName()),
			nullToEmpty(entity.getAccountNumber()),
			nullToEmpty(entity.getAccountHolder())
		);
	}

	private String toBankName(BankName bankName) {
		return bankName == null ? BankName.NONE.getDisplayName() : bankName.getDisplayName();
	}

	private String nullToEmpty(String value) {
		return value == null ? "" : value;
	}
}

File: infra/src/main/kotlin/com/beat/infra/persistence/booking/repository/query/MakerTicketReadPortImpl.kt
package com.beat.infra.persistence.booking.repository.query

import com.beat.contracts.booking.MakerTicketReadPort
import com.beat.contracts.booking.readmodel.MakerTicketBookingStatus
import com.beat.contracts.booking.readmodel.MakerTicketListItemReadModel
import com.beat.contracts.booking.readmodel.MakerTicketScheduleNumber
import com.beat.domain.booking.domain.BookingStatus
import com.beat.domain.performance.domain.BankName
import com.beat.domain.schedule.domain.ScheduleNumber
import com.beat.infra.persistence.booking.entity.BookingJpaEntity
import com.beat.infra.persistence.schedule.entity.ScheduleJpaEntity
import com.linecorp.kotlinjdsl.dsl.jpql.jpql
import com.linecorp.kotlinjdsl.querymodel.jpql.select.SelectQuery
import com.linecorp.kotlinjdsl.render.jpql.JpqlRenderContext
import com.linecorp.kotlinjdsl.support.hibernate.extension.createQuery
import jakarta.persistence.EntityManager
import org.springframework.stereotype.Repository

@Repository
// JpqlRenderContext is provided by JpaConfig, which is wired via @EnableInfraBaseConfig's
// DeferredImportSelector. IntelliJ cannot statically trace that path, so the injection is flagged as a
// false positive; runtime wiring is verified by the module context-boot integration tests.
@Suppress("SpringJavaInjectionPointsAutowiringInspection")
class MakerTicketReadPortImpl(
    private val entityManager: EntityManager,
    private val jpqlRenderContext: JpqlRenderContext,
) : MakerTicketReadPort {

    override fun findTickets(
        performanceId: Long?,
        scheduleNumbers: List<MakerTicketScheduleNumber>,
        bookingStatuses: List<MakerTicketBookingStatus>,
    ): List<MakerTicketListItemReadModel> {
        val query = buildTicketQuery(performanceId, scheduleNumbers, bookingStatuses, searchWord = null)
        return entityManager.createQuery(query, jpqlRenderContext).resultList.map(::toReadModel)
    }

    override fun searchTickets(
        performanceId: Long?,
        searchWord: String?,
        scheduleNumbers: List<MakerTicketScheduleNumber>,
        bookingStatuses: List<MakerTicketBookingStatus>,
    ): List<MakerTicketListItemReadModel> {
        if (searchWord.isNullOrBlank()) {
            return emptyList()
        }
        val query = buildTicketQuery(performanceId, scheduleNumbers, bookingStatuses, searchWord)
        return entityManager.createQuery(query, jpqlRenderContext).resultList.map(::toReadModel)
    }

    private fun buildTicketQuery(
        performanceId: Long?,
        scheduleNumberNames: List<MakerTicketScheduleNumber>,
        bookingStatusNames: List<MakerTicketBookingStatus>,
        searchWord: String?,
    ): SelectQuery<BookingJpaEntity> {
        val scheduleNumbers = scheduleNumberNames.map(::toScheduleNumber)
        val bookingStatuses = bookingStatusNames.map(::toBookingStatus)

        return jpql {
            select(
                entity(BookingJpaEntity::class),
            ).from(
                entity(BookingJpaEntity::class),
                entity(ScheduleJpaEntity::class),
            ).whereAnd(
                path(BookingJpaEntity::scheduleId).eq(path(ScheduleJpaEntity::id)),
                path(BookingJpaEntity::bookingStatus).ne(BookingStatus.BOOKING_DELETED),
                performanceId?.let { path(ScheduleJpaEntity::performanceId).eq(it) },
                scheduleNumbers.takeIf { it.isNotEmpty() }
                    ?.let { path(ScheduleJpaEntity::scheduleNumber).`in`(it) },
                bookingStatuses.takeIf { it.isNotEmpty() }
                    ?.let { path(BookingJpaEntity::bookingStatus).`in`(it) },
                // MySQL full-text search: function('match', bookerName, :searchWord) > 0
                searchWord?.let {
                    function(Double::class, "match", path(BookingJpaEntity::bookerName), value(it)).gt(0.0)
                },
            ).orderBy(
                // 상태 우선순위 -> 최신순. 기존 JPQL ORDER BY CASE ... END ASC, createdAt DESC 와 동일.
                caseWhen(path(BookingJpaEntity::bookingStatus).eq(BookingStatus.REFUND_REQUESTED)).then(1)
                    .`when`(path(BookingJpaEntity::bookingStatus).eq(BookingStatus.CHECKING_PAYMENT)).then(2)
                    .`when`(path(BookingJpaEntity::bookingStatus).eq(BookingStatus.BOOKING_CONFIRMED)).then(3)
                    .`when`(path(BookingJpaEntity::bookingStatus).eq(BookingStatus.BOOKING_CANCELLED)).then(4)
                    .`else`(5)
                    .asc(),
                path(BookingJpaEntity::createdAt).desc(),
            )
        }
    }

    private fun toReadModel(entity: BookingJpaEntity): MakerTicketListItemReadModel =
        MakerTicketListItemReadModel(
            entity.id,
            entity.bookerName,
            entity.bookerPhoneNumber,
            entity.scheduleId,
            entity.purchaseTicketCount,
            entity.createdAt,
            toMakerTicketBookingStatus(entity.bookingStatus),
            toBankName(entity.bankName),
            entity.accountNumber ?: "",
            entity.accountHolder ?: "",
        )

    private fun toBankName(bankName: BankName?): String =
        (bankName ?: BankName.NONE).displayName

    private fun toScheduleNumber(scheduleNumber: MakerTicketScheduleNumber): ScheduleNumber =
        when (scheduleNumber) {
            MakerTicketScheduleNumber.FIRST -> ScheduleNumber.FIRST
            MakerTicketScheduleNumber.SECOND -> ScheduleNumber.SECOND
            MakerTicketScheduleNumber.THIRD -> ScheduleNumber.THIRD
            MakerTicketScheduleNumber.FOURTH -> ScheduleNumber.FOURTH
            MakerTicketScheduleNumber.FIFTH -> ScheduleNumber.FIFTH
            MakerTicketScheduleNumber.SIXTH -> ScheduleNumber.SIXTH
            MakerTicketScheduleNumber.SEVENTH -> ScheduleNumber.SEVENTH
            MakerTicketScheduleNumber.EIGHTH -> ScheduleNumber.EIGHTH
            MakerTicketScheduleNumber.NINTH -> ScheduleNumber.NINTH
            MakerTicketScheduleNumber.TENTH -> ScheduleNumber.TENTH
        }

    private fun toBookingStatus(bookingStatus: MakerTicketBookingStatus): BookingStatus =
        when (bookingStatus) {
            MakerTicketBookingStatus.CHECKING_PAYMENT -> BookingStatus.CHECKING_PAYMENT
            MakerTicketBookingStatus.BOOKING_CONFIRMED -> BookingStatus.BOOKING_CONFIRMED
            MakerTicketBookingStatus.BOOKING_CANCELLED -> BookingStatus.BOOKING_CANCELLED
            MakerTicketBookingStatus.REFUND_REQUESTED -> BookingStatus.REFUND_REQUESTED
            MakerTicketBookingStatus.BOOKING_DELETED -> BookingStatus.BOOKING_DELETED
        }

    private fun toMakerTicketBookingStatus(bookingStatus: BookingStatus): MakerTicketBookingStatus =
        when (bookingStatus) {
            BookingStatus.CHECKING_PAYMENT -> MakerTicketBookingStatus.CHECKING_PAYMENT
            BookingStatus.BOOKING_CONFIRMED -> MakerTicketBookingStatus.BOOKING_CONFIRMED
            BookingStatus.BOOKING_CANCELLED -> MakerTicketBookingStatus.BOOKING_CANCELLED
            BookingStatus.REFUND_REQUESTED -> MakerTicketBookingStatus.REFUND_REQUESTED
            BookingStatus.BOOKING_DELETED -> MakerTicketBookingStatus.BOOKING_DELETED
        }
}
